temporian/implementation/numpy/operators/assign.py

```diff
@@ -0,0 +1,200 @@
+    values where sampling_1 and sampling_2 have the same timestamp.
+
+    Args:
+        sampling_1 (NumpySampling): sampling of assignee event.
+        sampling_2 (NumpySampling): sampling of assigned event.
+        index (tuple): index to check.
+
+    Returns:
+        np.ndarray: boolean np.darray with same shape as sampling_1.data[index]
+
+    Examples:
+        >>> sampling_1 = NumpySampling(
+        ...     names=["user_id"],
+        ...     data={
+        ...         (151591562,): np.array(
+        ...             ["2022-02-05", "2022-02-06", "2022-02-07"],
+        ...             dtype="datetime64",
+        ...         ),
+        ...         }
+        ...     )
+        >>> sampling_2 = NumpySampling(
+        ...     names=["user_id"],
+        ...     data={
+        ...         (151591562,): np.array(
+        ...             ["2022-02-06", "2022-02-07", "2022-02-08"],
+        ...             dtype="datetime64",
+        ...         ),
+        ...         }
+        ...     )
+        >>> _get_common_timestamps(sampling_1, sampling_2, (151591562,))
+        array([False, True, True])
+
+    """
+    return np.in1d(sampling_1.data[index], sampling_2.data[index])
+
+
+def _convert_assigned_feature_to_assignee_sampling(
+    feature: NumpyFeature,
+    common_timestamps: List[bool],
+    assignee_sampling: NumpySampling,
+    assigned_sampling: NumpySampling,
+    index: tuple,
+) -> NumpyFeature:
+    """Converts assigned feature to assignee sampling. It requires a boolean list with
+    same length as the assignee sampling, where True values indicate that the
+    assigned feature has the same timestamp.
+
+    Args:
+        feature (NumpyFeature): assigned feature to convert.
+        common_timestamps (List[bool]): list of booleans indicating where feature's sampling has same timestamp as new sampling
+        assignee_sampling (NumpySampling): sampling of assignee event.
+        assigned_sampling (NumpySampling): sampling of assigned event.
+        index (tuple): index to check.
+
+    Returns:
+        NumpyFeature: feature with new sampling.
+
+    """
+    new_feature = NumpyFeature(
+        name=feature.name,
+        data=np.full(len(common_timestamps), np.nan),
+    )
+
+    # Loop over common timestamps and if True, then add the assigned feature value
+    # in the common timestamp.
+    # Assuming that the feature data is sorted by timestamp, we can use binary search
+    # to find the index of the common timestamp in the feature data, instead of
+    # numpy where search. This would make the algorithm O(n * (1 + log(m)))
+    # (n: len(common_timestamps), m: len(feature.data). Adding extra n for
+    # np.full at the beginning.
+    # We can make this more efficient by trimming the feature data to the
+    # last found array index because as it sorted by timestamps, the next one
+    # will be after that one.
+
+    for i, is_common in enumerate(common_timestamps):
```

Comment:
I think we can circumvent the `for` loop here and replace it with a vectorized implementation.
```
mask = (
        assignee_sampling.data[index][:, np.newaxis]
        == assigned_sampling.data[index][np.newaxis, :]
)
mask_i, mask_j = mask.nonzero()
new_feature.data[mask_i] = feature.data[mask_j]
```
This implementation also avoids using the `np.in1d` in `_get_common_timestamps`. Note: this only works if the assignee event does not contain repeated timestamps. We need to make some changes to the code to support that case.